ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/JdbcLedgerDao.scala

```diff
@@ -1610,6 +1611,115 @@ private class JdbcLedgerDao(
+    """.stripMargin)
+
+  case class ParsedCommandData(
+      submittedAt: Instant,
+      ttl: Instant,
+      success: Option[Boolean],
+      error: Option[String])
+
+  private val CommandDataParser: RowParser[ParsedCommandData] =
+    Macro.parser[ParsedCommandData](
+      "submitted_at",
+      "ttl",
+      "success",
+      "error"
+    )
+
+  override def deduplicateCommand(
+      deduplicationKey: String,
+      submittedAt: Instant,
+      ttl: Instant): Future[Option[CommandDeduplicationEntry]] =
+    dbDispatcher.executeSql("deduplicate_command") { implicit conn =>
+      // Remove expired deduplication rows
+      SQL_REMOVE_OLD_COMMAND
+        .on("deduplicationKey" -> deduplicationKey, "submittedAt" -> submittedAt)
+        .execute()
+
+      // Try to insert a new deduplication row.
+      // If the row already exists and the insert fails, select the existing row.
+      Try(
+        SQL_INSERT_COMMAND
+          .on("deduplicationKey" -> deduplicationKey, "submittedAt" -> submittedAt, "ttl" -> ttl)
+          .execute()
+      ).fold(
+        _ =>
+          Try(
+            SQL_SELECT_COMMAND
+              .on("deduplicationKey" -> deduplicationKey, "submittedAt" -> submittedAt)
+              .as(CommandDataParser.single)
+          ).fold(
+            error => sys.error(s"Could not find deduplication data: $error"), {
+              case ParsedCommandData(originalSubmittedAt, originalTtl, None, None) =>
+                Some(
+                  CommandDeduplicationEntry(
+                    deduplicationKey,
+                    originalSubmittedAt,
+                    originalTtl,
+                    None))
+              case ParsedCommandData(originalSubmittedAt, originalTtl, Some(true), None) =>
+                Some(
+                  CommandDeduplicationEntry(
+                    deduplicationKey,
+                    originalSubmittedAt,
+                    originalTtl,
+                    Some(Right(()))))
+              case ParsedCommandData(originalSubmittedAt, originalTtl, Some(false), Some(error)) =>
+                Some(
+                  CommandDeduplicationEntry(
+                    deduplicationKey,
+                    originalSubmittedAt,
+                    originalTtl,
+                    Some(Left(error))))
+              case data =>
+                sys.error(s"Invalid command deduplication row $data")
+            }
+        ),
+        _ => None
+      )
+    }
+
+  private val SQL_UPDATE_COMMAND = SQL(
+    """
+      |update participant_command_submissions
+      |set success={success}, error={error}
+      |where deduplication_key={deduplicationKey} and submitted_at={submittedAt}
+    """.stripMargin)
+
+  override def updateCommandResult(
+      deduplicationKey: String,
+      submittedAt: Instant,
+      result: Either[String, Unit]): Future[Unit] =
+    dbDispatcher.executeSql("deduplicate_command") { implicit conn =>
```

Comment:
```suggestion
    dbDispatcher.executeSql("update_command_result") { implicit conn =>
```
Not sure about the name, but it's probably better to not have two queries share the same key for metrics registering purposes.